File: packages/template-python/src/deepagents_template/runtime/config/config_merge.py
"""Config merge primitives — layered merge with array-concat semantics."""

from __future__ import annotations

from typing import Any

from deepagents_template.runtime.config.deep_merge import deepMerge


def setNestedValue(obj: dict[str, Any], path: str, value: Any) -> None:
    keys = path.split(".")
    current = obj
    for key in keys[:-1]:
        if key not in current or current[key] is None or not isinstance(current[key], dict):
            current[key] = {}
        current = current[key]
    current[keys[-1]] = value


def concatUnique(a: list[str], b: list[str]) -> list[str]:
    seen = set(a)
    result = list(a)
    for item in b:
        if item not in seen:
            seen.add(item)
            result.append(item)
    return result


def isRecord(value: Any) -> bool:
    return isinstance(value, dict)


def mergeConfigLayer(config: dict[str, Any], layer: dict[str, Any]) -> dict[str, Any]:
    prev_skills = config.get("skills", {}).get("directories", [])
    prev_agents = config.get("agentsDirectories", [])
    prev_mcp_paths = config.get("mcp", {}).get("configPaths", [])
    prev_mcp_servers = config.get("mcp", {}).get("servers", {})
    prev_plugin_dirs = config.get("plugins", {}).get("directories", [])
    merged = deepMerge(config, layer)
    if "skills" in layer and "directories" in layer.get("skills", {}):
        merged["skills"]["directories"] = concatUnique(
            list(prev_skills), list(layer["skills"]["directories"])
        )
    if "agentsDirectories" in layer:
        merged["agentsDirectories"] = concatUnique(
            list(prev_agents), list(layer["agentsDirectories"])
        )
    if "mcp" in layer and "configPaths" in layer.get("mcp", {}):
        merged["mcp"]["configPaths"] = concatUnique(
            list(prev_mcp_paths), list(layer["mcp"]["configPaths"])
        )
    if "mcp" in layer and "servers" in layer.get("mcp", {}):
        merged["mcp"]["servers"] = {**prev_mcp_servers, **layer["mcp"]["servers"]}
    if "plugins" in layer and "directories" in layer.get("plugins", {}):
        merged["plugins"]["directories"] = concatUnique(
            list(prev_plugin_dirs), list(layer["plugins"]["directories"])
        )
    return merged

File: packages/deepagents-flow-ts/src/runtime/config/config-merge.ts
/**
 * Config merge primitives.
 *
 * Layered merge with array-concat semantics for the fields where a later layer
 * should add to (not replace) the earlier one. Extracted from config-loader.ts.
 */
import type { AppConfig } from "./config-schema.js";
import { deepMerge } from "./deep-merge.js";

/** Set a nested value in an object using a dot-separated path */
export function setNestedValue(obj: Record<string, unknown>, path: string, value: unknown): void {
  const keys = path.split(".");
  let current = obj;
  for (let i = 0; i < keys.length - 1; i++) {
    const key = keys[i]!;
    if (!(key in current) || current[key] === null || typeof current[key] !== "object") {
      current[key] = {};
    }
    current = current[key] as Record<string, unknown>;
  }
  current[keys[keys.length - 1]!] = value;
}

export function concatUnique(a: string[], b: string[]): string[] {
  return Array.from(new Set([...a, ...b]));
}

export function isRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === "object" && !Array.isArray(value);
}

export function mergeConfigLayer(config: AppConfig, layer: Partial<AppConfig>): AppConfig {
  const previousSkills = config.skills.directories;
  const previousAgents = config.agentsDirectories;
  const previousMcpPaths = config.mcp.configPaths;
  const previousMcpServers = config.mcp.servers;
  const previousPluginDirs = config.plugins.directories;
  const merged = deepMerge(config, layer);

  if (layer.skills?.directories) {
    merged.skills.directories = concatUnique(previousSkills, layer.skills.directories);
  }
  if (layer.agentsDirectories) {
    merged.agentsDirectories = concatUnique(previousAgents, layer.agentsDirectories);
  }
  if (layer.mcp?.configPaths) {
    merged.mcp.configPaths = concatUnique(previousMcpPaths, layer.mcp.configPaths);
  }
  if (layer.mcp?.servers) {
    merged.mcp.servers = { ...previousMcpServers, ...layer.mcp.servers };
  }
  if (layer.plugins?.directories) {
    merged.plugins.directories = concatUnique(previousPluginDirs, layer.plugins.directories);
  }

  return merged;
}
